Document users hashed passwords with argon2 does not work in FIPS env… (#33481)
keycloak/keycloak: docs/guides/server/fips.adoc

supported in {project_name} when using strict mode. Some examples are importing or generating a keystore of an OIDC or SAML client in the Admin Console or for a `java-keystore` provider in the realm keys.

* User passwords must be 14 characters or longer. {project_name} uses PBKDF2 based password encoding by default. BCFIPS approved mode requires passwords to be at least 112 bits
(effectively 14 characters) with PBKDF2 algorithm. If you want to allow a shorter password, set the property `max-padding-length` of provider `pbkdf2-sha512` of SPI `password-hashing`

> to value 14 to provide additional padding when verifying a hash created by this algorithm. This setting is also backwards compatible with previously stored passwords.

Reviewer: ```suggestion
to 14 to provide additional padding when verifying a hash created by this algorithm. This setting is also backwards compatible with previously stored passwords.
```

For example, if the user's database is in a non-FIPS environment and you have shorter passwords and you want to verify them now with {project_name} using BCFIPS in approved mode, the passwords should work.
So effectively, you can use an option such as the following when starting the server:

[source]
----
--spi-password-hashing-pbkdf2-sha512-max-padding-length=14
----

NOTE: Using the option above does not break FIPS compliance. However, note that longer passwords are good practice anyway. For example, passwords auto-generated by modern browsers match this

> requirement as they are longer than 14 characters. If you want to omit the option for max-padding-length, you can set password policy to your realms to have passwords at least 14 characters long.

Reviewer: ```suggestion
requirement as they are longer than 14 characters. If you want to omit the option for max-padding-length, you can set the password policy to your realms to have passwords at least 14 characters long.
```

> NOTE: When you are migrating from {project_name} older than 24, or if you explicitly set the password policy to override the default hashing algorithm, it is possible that some of your users use an older

Reviewer: ```suggestion
NOTE: When you are migrating from {project_name}  prior to version 24, or if you explicitly set the password policy to override the default hashing algorithm, it is possible that some of your users use an older
```

> NOTE: When you are migrating from {project_name} older than 24, or if you explicitly set the password policy to override the default hashing algorithm, it is possible that some of your users use the older

Reviewer 2: ```suggestion
NOTE: When you are migrating from {project_name} older than 24, or if you explicitly set the password policy to override the default hashing algorithm, it is possible that some of your users use an older
```

algorithm like `pbkdf2-sha256`. In this case, it may be good to add also the option `--spi-password-hashing-pbkdf2-sha256-max-padding-length=14` to make sure that users having their passwords hashed with

> older `pbkdf2-sha256` are able to login as well in case that their passwords are shorter than 14 characters.

Reviewer 2: ```suggestion
the older `pbkdf2-sha256` are able to login as well in case that their passwords are shorter than 14 characters.
```

> algorithm like `pbkdf2-sha256`. In this case, it may be good to add also the option `--spi-password-hashing-pbkdf2-sha256-max-padding-length=14` to make sure that users having their passwords hashed with
> the older `pbkdf2-sha256` are able to login as well in case that their passwords are shorter than 14 characters.

Reviewer: ```suggestion
algorithm like `pbkdf2-sha256`. In this case, consider adding  the `--spi-password-hashing-pbkdf2-sha256-max-padding-length=14` option to ensure that users having their passwords hashed with
the older `pbkdf2-sha256` can log in because their passwords may be shorter than 14 characters.
```

If you previously used {project_name} in a non-fips environment, it is possible to migrate it to a FIPS environment including its data. However, restrictions and considerations exist as
mentioned in previous sections, namely:

> * From {project_name} 25, the default algorithm for password hashing is `argon2`. However this algorithm is not supported for FIPS 140-2. This means that if your users hashed their password
> with `argon2`, they will not be able to login after switch to the FIPS environment. Ideally, if you are aware that you will need to migrate to the FIPS environment, it can be good to set the

Reviewer: ```suggestion
* Starting with {project_name} 25, the default algorithm for password hashing is `argon2`. However, this algorithm is not supported for FIPS 140-2. As a result, if users hashed their password
with `argon2`, they can log in after switching to the FIPS environment. Therefore, if you plan  to migrate to the FIPS environment, consider setting the 
```

Password policy for your realm from the beginning (before any users are created) and override the default algorithm for example to `pbkdf2-sha512`, which is FIPS compliant. This means that
migration to the FIPS environment will be smooth. Otherwise, if your users are already on `argon2` passwords, it may be needed to ask your users to reset password after migrating to the FIPS

> environment - for instance by asking to use "Forget password" functionality or by sending the email for reset-password to all the users.

Reviewer: ```suggestion
environment.  For instance, ask users to use "Forget password" or send the email for reset-password to all users.
```
